Update Operator-Status enhancement (#51)
operator-framework/enhancements: enhancements/operator-conditions.md

## Design Details

### status.operators.coreos.com CRD

OLM will create a channel for operators to communicate complex conditions by introducing the `Status` [CustomResourceDefinition](https://kubernetes.io/docs/concepts/extend-kubernetes/api-extension/custom-resources/).

> When an operator is installed, OLM will create a `Status` CustomResource (CR) for the operator. The operator's RBAC will be configured such that it is only able to get the `Status` resource it own and update the resources [status subresource](https://book-v1.book.kubebuilder.io/basics/status_subresource.html).

Reviewer: IMO, `Status` is even more confusing that `Conditions` b/c of the status subresource. Perhaps its best for us to get away from overloaded terms, maybe `Probe` is more apt (slightly less overloaded).

Author: I don't believe that probe is appropriate given that the operator is basically communicating its status to OLM via this resource (I am purposely going to advocate that we avoid discussing the features in #23 that led to this name, but if we do add these features in the future we could do so with a spec.probe field).

Author: There was also concern that the `Conditions` name is just as overloaded as `Status` and limits the CRD from expanding to additional usecases outside of communicating state through the `status.conditions` array.

When the `Status` CR is created, OLM will update the deployments defined in the CSV with two addition environment variables:

> - *OperatorNamespace*: The namespace that the operator is running in

Reviewer: I don't think OLM needs to add this, since it's a standard use case for the [downward API](https://kubernetes.io/docs/tasks/inject-data-application/environment-variable-expose-pod-information/#use-pod-fields-as-values-for-environment-variables).

Author: Pods are not automatically generated with this information - are you suggesting that we instruct the sdk to always include the field reference?

Author: If so - that seems fine to me.

CC @varshaprasad96 @joelanford please comment if you object.

Reviewer: > are you suggesting that we instruct the sdk to always include the field reference?

No, just that the operator should know what namespace it's in, one way or another -- that environment variable can be added manually.

- *OperatorStatusName*: The name of the operator's `Status` CR

> Operators will then be use the name and namespace of its `Status` CR along with a library provided by the [operator-sdk](https://github.com/operator-framework/operator-sdk) team to set [conditions](https://github.com/kubernetes/enhancements/blob/3aa92e20bdfa2e60e3cb1a2b92cdca61847d7ad2/keps/sig-api-machinery/1623-standardize-conditions/README.md#kep-1623-standardize-conditions) on the CR to communicate complex states to OLM. For those unfamiliar with conditions, they are used by many Kubernetes Resources to provide users with additional insight into the state of the resource and typically adhere to the following format:

Reviewer: ```suggestion
With the aid of a library provided by the [operator-sdk](https://github.com/operator-framework/operator-sdk), operators may use to set [conditions](https://github.com/kubernetes/enhancements/blob/3aa92e20bdfa2e60e3cb1a2b92cdca61847d7ad2/keps/sig-api-machinery/1623-standardize-conditions/README.md#kep-1623-standardize-conditions) on the respective `Status` CR to communicate complex states to OLM. For those unfamiliar with conditions, they are used by many Kubernetes Resources to provide users with additional insight into the state of the resource and typically adhere to the following format:
```

The `Upgradeable` "OLM Supported Condition" allows an operator to communicate when OLM should or should not upgrade the operator.

When the `Upgradeable` condition is set to `False`, OLM will:

- Not perform automatic or manual updates to the operator.
- Not update the operator as a part of [dependency resolution](./operator-dependency-resolution.md).

> By default, OLM prevents upgrades to operators when the CSV is not in the `Succeeded` phase. When the `Upgradeable` condition is set to `True`, OLM will not prevent upgrades to the operator.

Reviewer: ```suggestion
By default, OLM prevents upgrades to operators when the CSV is not in the `Succeeded` phase. When the `Upgradeable` condition is set to `True`, OLM will allow upgrades to the operator.
```

- An operator is about to start a critical process and should not be upgraded until after the process is completed.
- The operator is performing a migration of CRs that must be completed before the operator is ready to be upgraded.

#### Ready

> The `Ready` "OLM Supported Condition" allows an operator to communicate its readiness to OLM. By default, OLM will move a CSV in the `Succeeded` Phase to the `Failed` phase if the deployments [Readiness Probe](https://kubernetes.io/docs/tasks/configure-pod-container/configure-liveness-readiness-startup-probes/#define-readiness-probes) is not true. By setting the `Ready` Condition, OLM will:

Reviewer 2: Is there something missing after the last line `By setting the Ready Condition, OLM will:`?

Author: Leftover copy-paste, I'll remove this.

The `Ready` condition might be useful when:

> - An operator wants to change its Readiness Probe status to 'false' to communicate information to a cluster admin but does not want OLM to move the operator into the failed state.

Reviewer 3: What OLM behavior changes if `Ready=true` vs. `Ready=false`?

I understood "supported conditions" to be things that would affect OLM behavior (i.e. Upgradeable=false prevents upgrade)

That's not to say an author couldn't write a `Ready` status. But it's more likely useful for the operator to communicate more detailed status via its own CRs (i.e. connection state to cloud accounts).

Author: Evan and I discussed this condition offline and ultimately reached the conclusion that this condition should be removed.

> - An operator wants to change it's Readiness Probe status to 'false' to communicate information to a cluster admin but does not want OLM to move the operator into the failed state.

Reviewer 2: Nit:
```suggestion
- An operator wants to change its Readiness Probe status to `false` to communicate information to a cluster admin but does not want OLM to move the operator into the failed state.
```
